src/lib/reasoners/uf.mli

```diff
@@ -34,6 +34,77 @@ type t
 
 type r = Shostak.Combine.r
 
+type _ id = ..
+(** Extensible type for global domains identifiers, see {!Domains}. *)
+
+module type Domains = sig
+  (** Module signature for global domains used by the union-find module.
+
+      The union-find module updates global domains when equivalence classes are
+      merged. *)
+
+  type t
+  (** The type of global domains. *)
+
+  val pp : t Fmt.t
+  (** Pretty-printer for global domains. *)
+
+  type _ id += Id : t id
+  (** Id used by the union-find module to dispatch to this module. Must be
+      unique (i.e. do not re-export the [Id] from another module). *)
+
+  val empty : t
+  (** The empty domain. *)
+
+  val filter_ty : Ty.t -> bool
+  (** Limit the type of terms that this module cares about. Only substitutions
+      of representatives for which [filter_ty (type_info r)] holds will be
+      propagated to this module. *)
+
+  val add : r -> t -> t
+  (** [add r t] is called when the representative [r] is added to the
+      union-find, if it has a type that matches [filter_ty].
+
+      {b Note}: unlike [Relation.add], this function is called even for
+      "dynamic" representatives added by [X.make] or AC normalization. *)
+
+  exception Inconsistent of Explanation.t
+  (** Raised by [subst] below when an inconsistency is found while merging the
+      domains.*)
+
+  val subst : ex:Explanation.t -> r -> r -> t -> t
+  (** [subst ~ex rr nrr t] is called when the representatives [rr] and [nrr] are
+      merged with explanation [ex]. [nrr] is the new representative; [rr] should
+      no longer be tracked and the intersection of domains should be associated
+      with [nrr].
+
+      The explanation [ex] justifies the equality [rr = nrr].
+
+      @raise Inconsistent if the domains for [rr] and [nrr] are incompatible. *)
+end
+
+type 'a domains = (module Domains with type t = 'a)
+(** The type for global domain modules with storage type ['a]. *)
```

Comment:
It is a domain (i.e. set of possible values) for the tuple comprising all of the currently defined variables (hence "global"). It so happens that all the implementations we currently have are represented as a product of per-variable ("local") domains, but I would rather not enshrine that fact in the documentation since one of the motivations is to use this API for the "delta" domain of [Sharpening Constraint Programming approaches for Bit-Vector Theory](https://cea.hal.science/cea-01795779/document).

I can clarify the documentation to mention that a global domain encodes a set of (correlated) values for multiples variables.

Reply:
> I can clarify the documentation to mention that a global domain encodes a set of (correlated) values for multiples variables.

Yes, please ;)